src/Bicep.Core/LanguageConstants.cs

```diff
@@ -132,24 +139,41 @@ public static IEnumerable<TypeProperty> GetCommonResourceProperties(ResourceType
             yield return new TypeProperty("apiVersion", new StringLiteralType(reference.ApiVersion), TypePropertyFlags.ReadOnly | TypePropertyFlags.SkipInlining);
         }
 
-        public static TypeSymbol CreateModuleType(IEnumerable<TypeProperty> paramsProperties, IEnumerable<TypeProperty> outputProperties, string typeName)
+        private static ResourceScopeReference CreateResourceScopeReference(ResourceScopeType resourceScope)
+            => resourceScope switch {
+                ResourceScopeType.TenantScope => new ResourceScopeReference("tenant", resourceScope),
+                ResourceScopeType.ManagementGroupScope => new ResourceScopeReference("managementGroup", resourceScope),
+                ResourceScopeType.SubscriptionScope => new ResourceScopeReference("subscription", resourceScope),
+                ResourceScopeType.ResourceGroupScope => new ResourceScopeReference("resourceGroup", resourceScope),
+                _ => new ResourceScopeReference("none", ResourceScopeType.None),
+            };
+
+        public static TypeSymbol CreateModuleType(IEnumerable<TypeProperty> paramsProperties, IEnumerable<TypeProperty> outputProperties, ResourceScopeType moduleScope, ResourceScopeType containingScope, string typeName)
         {
             var paramsType = new NamedObjectType(ModuleParamsPropertyName, TypeSymbolValidationFlags.Default, paramsProperties, null);
             // If none of the params are reqired, we can allow the 'params' declaration to be ommitted entirely
             var paramsRequiredFlag = paramsProperties.Any(x => x.Flags.HasFlag(TypePropertyFlags.Required)) ? TypePropertyFlags.Required : TypePropertyFlags.None;
 
             var outputsType = new NamedObjectType(ModuleOutputsPropertyName, TypeSymbolValidationFlags.Default, outputProperties, null);
+            var resourceRefArray = new TypedArrayType(ResourceRef, TypeSymbolValidationFlags.Default);
 
-            return new NamedObjectType(
+            // If the module scope matches the parent scope, we can safely omit the scope property
+            var scopeRequiredFlag = moduleScope != containingScope ? TypePropertyFlags.Required : TypePropertyFlags.None;
+
+            var moduleBody = new NamedObjectType(
                 typeName,
                 TypeSymbolValidationFlags.Default,
                 new []
                 {
                     new TypeProperty(ModuleNamePropertyName, LanguageConstants.String, TypePropertyFlags.Required | TypePropertyFlags.SkipInlining),
+                    new TypeProperty(ModuleScopePropertyName, CreateResourceScopeReference(moduleScope), TypePropertyFlags.WriteOnly | scopeRequiredFlag),
```

Comment:
>scopeRequiredFlag [](start = 135, length = 17)

Shouldn't it be always optional, but always validated (taking defaults into account)?

Reply:
There are cases where it's non-optional - for example if invoking a resourceGroup-level module through a subscription-level parent file